README.md
# Vacation Soup Stock Installer
[![GitHub license](https://img.shields.io/badge/license-GPLv3-blue.svg)](https://raw.githubusercontent.com/francisw/soup-waiter/release/LICENSE)


- **Lead Developer:**
[Francis Wallinger](https://github.com/francisw)
- **Version:** [0.1.0](https://github.com/francisw/soup-stock)
- **Requires at least:** 4.7.1 
- **Tested up to:** 4.8.2

## Description

Soup Stock is a tool to install a new copy of stock

## Installation

This is a standard Wordpress plugin, and is installed using the wp-admin plugin installer.

1. Install a new clean copy of Wordpress. Do not do anything else to it except this.
2. Download the latest Source code(zip) from our [Github](https://github.com/francisw/soup-release/release) repository
3. In the wp-admin dashoard panel select Plugins->Add New
4. Click Upload and select the zip file you downloaded for this plugin
5. Click Activate - Your Wordpress installation will now be replaced with Vacation Soup Stock


## Usage

This is an unattended plugin. Just activate it and it will install the Vacation Soup Stock. 

When it is complete, your page will automatically refresh with your new site. 
Log in (add /wp-admin to the url) with the credentials you used to create the Wordpress website, and start building your site.


## Common problems

This only works on a clean, brand new installation of Wordpress. If you have done anything in Wordpress other than install this plugin, it will not run in order to protect what you have done. To correct this, delete your Wordpress installation and start again.

## Feedback

See [our community](https://community.vacationsoup.com) for details on how to participate. If you are interested in contributing code to the project, please submit pull requests or [contact me](mailto://francis@vacationsoup.com) directly.

## Changelog

See [CHANGELOG.md](CHANGELOG.md).



CHANGELOG.md
# Changelog for Soup Stock

## 0.1.0 (2017-10-14)
* **Enhancements**:
 - Running tests remotely

## 0.0.1 (2017-10-12)
* **Enhancements**:
 - Created plugin
 - Running tests locally


Stock/Base.php
<?php
namespace Stock;

use Exception;
/**
 * Created by PhpStorm.
 * User: francisw
 * Date: 12/09/2017
 * Time: 11:31
 */
class Base {
	const PFX_GET = 'get_';
	const PFX_SET = 'set_';
	const PFX_IS = 'is_';

	/**
	 * To be extended by inheritors if persistence needed, as in Singleton
	 */
	protected function persist($key=null){
	}
	static protected function getPersisted($key=null){
		return null;
	}

	public function __set($name,$value){
		$method = self::PFX_SET.$name;
		if(method_exists($this,$method)){
			$this->$method($value);
		} else {
			// Getter/Setter not defined so set as property of object
			if (property_exists($this,$name)) {
				$this->$name = $value;
			} else {
				throw new Exception("Property '{$name}' does not exist in ".get_called_class());
			}
		}
		$this->persist();
	}

	public function __get($name){
		$methods = [self::PFX_GET,self::PFX_IS];
		foreach ($methods as $method){
			$method .= $name;
			if(method_exists($this, $method)){
				return $this->$method();
			}
		}
		if (property_exists($this,$name)) {
			// Getter/Setter not defined so return property if it exists
			return $this->$name;
		} else {
			throw new Exception("Property '{$name}' does not exist in ".get_called_class());
		}
	}

	public function __isset($name){
		return (
			property_exists($this,$name) ||
			method_exists($this,self::PFX_SET.$name) || // For a write-only property
			method_exists($this,self::PFX_GET.$name) || // For a read-only property
			method_exists($this,self::PFX_IS.$name)     // For a read-only boolean property
		);
	}
	public function __unset($name){
		unset($this->$name);
	}
}

soup-stock.php
<?php
namespace Stock;

/**
 * The plugin bootstrap file
 *
 * This file is read by WordPress to generate the plugin information in the
 * plugin admin area. This file also includes all of the dependencies used by
 * the plugin, registers the activation and deactivation functions, and defines
 * a function that starts the plugin.p
 *
 * @link              https://github.com/francisw/soup-release
 * @since             4.8.1
 * @package           soup_release
 *
 * @wordpress-plugin
 * Plugin Name:       Vacation Soup Stock Installer
 * Plugin URI:        https://github.com/francisw/soup-release/release
 * Description:       Backup Stock for Distribution
 * Version:           0.1.0
 * Author:            Francis Wallinger
 * Author URI:        http://github.com/francisw
 * License:           GPL-3.0+
 * License URI:       http://www.gnu.org/licenses/gpl-3.0.txt
 */
// If this file is accessed directory, then abort.
if ( ! defined( 'WPINC' ) ) {
	throw new \Exception("Cannot be accessed directly");
}
if ( ! function_exists( 'version_compare' ) || version_compare( PHP_VERSION, '5.3.0', '<' ) ) {
	throw new \Exception("PHP Version not supported");
}

// Now check he required plugins

require_once("inc/autoloader.php");

add_action("admin_init",array(SoupStock::single(),'init'));


inc/autoloader.php
<?php
/**
 * Copied from: https://github.com/php-fig/fig-standards/blob/master/accepted/PSR-4-autoloader-examples.md
 **
 * @param string $class The partially-qualified class name.
 * @return void
 */
spl_autoload_register(function ($class) {

	// base directory for the namespace prefix
	$base_dir = __DIR__ . '/../';

	// project-specific namespace prefix
	$prefix ='Stock\\';

	// does the class use the namespace prefix?
	$len = strlen($prefix);
	if (strncmp($prefix, $class, $len) !== 0) {
		// no, move to the next registered autoloader
		return;
	}

	/* // get the relative class name
	$relative_class = substr($class, $len*/

	// replace the namespace prefix with the base directory, replace namespace
	// separators with directory separators in the relative class name, append
	// with .php
	$file = $base_dir . str_replace('\\', '/', $class) . '.php';

	// if the file exists, require it
	if (file_exists($file)) {
		require $file;
	}
});

Stock/SoupRelease.php
<?php
/**
 * Created by PhpStorm.
 * User: francisw
 * Date: 11/10/2017
 * Time: 11:13
 */

namespace Stock;


class SoupRelease extends SitePersistedSingleton
{
    protected $last_run_at;
    protected $last_result;
    protected $last_eresult;
    protected $state;

    public function init(){
        switch ($this->state){
            case 'persisted':
                // We have been restored from a database, so we need to
                // Deactivate ourselves
                deactivate_plugins( plugin_basename( __DIR__ ) );
                $this->soup_release_deactivation(); // Make sure
                break;
            default:
                $this->state='releasing';
                $this->persist();
            case 'releasing':
                if ( current_user_can( 'manage_options' ) ) {
                    if ( ! wp_next_scheduled( 'soup_release_hook' ) ) {
                        wp_schedule_event( time(), 'daily', 'soup_release_hook' );
                        register_deactivation_hook(__FILE__, [$this,'soup_release_deactivation']);
                    }
                    add_action( 'soup_release_hook', [$this,'soup_release_action'] );
                    $this->_release();// Enable release from command line to push an urgent change, append &_release to URL
                    // $this->_extend();// Enable release from command line to push an urgent change, append &_release to URL
                }
        }
    }

    function soup_release_action() {
        GLOBAL $table_prefix;
        $server_name   = DB_HOST;
        $username      = DB_USER;
        $password      = DB_PASSWORD;
        $database_name = DB_NAME;
        $backup_path   = __DIR__ . '/../current';

        // Set our state so a restored copy will know it should die and disable
        $this->state='persisted';
        $this->persist();

        $prog          = 'bin/fail';
        foreach (['/usr/local/bin/mysqldump','/usr/bin/mysqldump','/Applications/MAMP/Library/bin/mysqldump'] as $prog){
            if (file_exists($prog)) {
                break;
            }
        }
        $mysqlparams = " -h {$server_name} -u {$username} \"-p{$password}\" {$database_name}";
        $mysqldump   = "{$prog} --hex-blob --routines --skip-lock-tables"; // --log-error={$backup_path}/mysqldump_error.log";

        $this->last_run_at = time();
        $this->last_result=$this->exec_array([
            "/bin/rm -f {$backup_path}/stock.*",
            "{$mysqldump} {$mysqlparams} | sed -e 's/{$table_prefix}/vsoup_/g' > {$backup_path}/stock.sql ",
            "tar -cf - --exclude='stock.tar*' . | /bin/gzip -c > {$backup_path}/stock.tar.gz",
        ]);
        $this->state='releasing'; // Reset our state so we know we are still running
        $this->persist(); // https://freevacationrentalwebsite.com/wp-content/plugins/soup-release-master/current/stock.tar.gz
    }
    function soup_extend_action() {

        $this->state='releasing';
        $this->persist();

        $this->soup_release_action();
    }

    function exec_array($cmds) {
        $result=[];
        foreach ($cmds as $command){
            $return = $output = 'Not run';
            $when = date("D M j G:i:s T Y");
            $ex = 'cd '.ABSPATH.' && '.$command.' 2>&1';
            exec($ex, $output, $return);
            $result[] = compact('when','command','ex','output','return');
            if (0!=$return){ // Oops
                break;
            }
        }
        return $result;
    }


    private function _release(){
        $last = 'last';
        if (isset($_REQUEST['_release'])){
            if ('last'!=$_REQUEST['_release']) {
                $this->state='releasing';
                $this->persist();
                $this->soup_release_action();
                $last = 'this';
            }
            echo "<h1>Output from {$last} run</h1><pre>";
            echo print_r($this->last_result);
            echo "</pre>";
            wp_die();
        }
    }
    private function _extend()
    {
        $last = 'last';
        if (isset($_REQUEST['_extend'])) {
            if ('last' != $_REQUEST['_extend']) {
                $this->soup_extend_action();
                $last = 'this';
            }
            echo "<h1>Output from {$last} run</h1><pre>";
            echo print_r($this->last_eresult);
            echo "</pre>";
            wp_die();
        }
    }

    function soup_release_deactivation() {
        wp_clear_scheduled_hook('soup_release_hook');
    }
}

soup-stock-installing.php
<?php
/**
 * Created by PhpStorm.
 * User: francisw
 * Date: 13/10/2017
 * Time: 12:25
 */

// Very basic, as Wordpress is being reinstalled while this is showing, so cannot rely on anything Pressy
?><!DOCTYPE html>
<!--[if IE 8]>
<html xmlns="http://www.w3.org/1999/xhtml" class="ie8 wp-toolbar"  lang="en-US">
<![endif]-->
<!--[if !(IE 8) ]><!-->
<html xmlns="http://www.w3.org/1999/xhtml" class="wp-toolbar"  lang="en-US">
<!--<![endif]-->
<head>
	<meta http-equiv="Content-Type" content="text/html; charset=UTF-8" />
	<title>Installing Vacation Soup Stock</title>
	<style type="text/css">
		body, html {
			height: 100%;
		}

		.bg {
			/* The image used */
			background-image: url("https://launch.vacationsoup.com/wp-content/uploads/2017/07/bigstock-177198292.jpg");

			/* Full height */
			height: 100%;

			/* Center and scale the image nicely */
			background-position: center;
			background-repeat: no-repeat;
			background-size: cover;

            font-family: "Open Sans", sans-serif;
		}
		#content {
            width: 60%;
            margin: 20% 0 0 20%;
            text-align: center;
        }
        #content IMG {
            display: block;
            min-width: 10%;
            margin: auto;
            fill: #8888ff;
        }
	</style>
</head>
<body class="bg">
<div id="content">
	<h1>Installing Vacation Soup Stock</h1>
    <img src="img/loading.svg" />
	<h3>You will be redirected to your login screen when this is complete.</h3>
</div>
</body>
</html>

Stock/SoupStock.php
<?php
/**
 * Created by PhpStorm.
 * User: francisw
 * Date: 12/10/2017
 * Time: 14:47
 */

namespace Stock;

/**
 * Class SoupStock
 * @package Stock
 *
 * This must only be installed on a virgin Wordpress installation, and it will replace it with
 * Vacation Soup's Stock version and database, and it will therefore no longer exist.
 */

class SoupStock extends Singleton
{
    const RELEASE_DOMAIN = "stock.vacationsoup.com";
    const RELEASE_RRL = "wp-content/plugins/soup-release-master/current";
    const RELEASE_PFX = "vsoup_";

    /**
     * @var string[] $results of commands run
     */
    protected $results;

    public function __construct(){
        $this->results=[];
    }

    public function init(){
    	global $table_prefix,$wpdb;
        // We can't have to follow normal Wordpress conventions here because we are
        // replacing the Wordpress installation with the Soup, so we run it all from within init


        try {
	        if (isset($_REQUEST['nosoup'])){
		        return;
	        }

	        // Set a splash screen
	        if (!isset($_REQUEST['_holding'])){
		        $this->showHoldingPage(); // Holding page creates request again
		        die();
	        }
	        $this->checkForNewInstallation();

	        // Firstly prevent multiple calls. Have observed WP calling activation 3 times
	        if (true===get_option('vs-stock-installing')){
		        throw new \Exception("Already installing Vacation Soup stock");
	        }
	        update_option('vs-stock-installing',true);

	        $stock='stock.tar';
	        //$file_url="https://".self::RELEASE_DOMAIN."/".self::RELEASE_RRL."/{$stock}.gz";
	        $file_url="https://github.com/francisw/stock/blob/master/stock.tar.gz?raw=true";
            $old_table_prefix = $table_prefix;
            $new_table_prefix = 'vsoup_';
            $sql_foreign = ABSPATH.self::RELEASE_RRL.'/'.'stock.sql';
            $sql_local = "/tmp/stock.sql";
            if (true) { //(@copy($file_url, $soup)) { //Grab the latest archive <-- now doing it from commands below

	            set_time_limit(30);
                if (TRUE || $_SERVER['HTTP_HOST'] != self::RELEASE_DOMAIN) {
                    // Save the wp-config, we will only change the $table_prefix
                    // Save the logged in User record, incl Password etc
                    // Save site urls (possibly, depending if dom

                    // OK, we're up - NO MORE wp commands from here, only raw php / unix, as we are replacing WP
                    try {
                    	$this->exec_array([
	                        //  move our website to a safe folder
                            "mkdir pre-vs",
                            "mv *  pre-vs || true", // To allow the error on moving pre-vs to itself
		                    // Get the stock
		                    "curl -Lo stock.tar.gz {$file_url}",
                            // Unpack the stock into this location
		                    "/bin/gzip -d {$stock}",
		                    "tar -xf {$stock}",
                            // Install the saved wp-config, changing the table prefix
                            "sed -e 's|\$table_prefix|\$table_prefix = \"vsoup_\"; // \$table_prefix|' < pre-vs/wp-config.php > wp-config.php",
		                    // And change any VacationSoup urls to local ones in the SQL
							"sed -e 's|".self::RELEASE_DOMAIN."|".$_SERVER['HTTP_HOST']."|g' < {$sql_foreign} > {$sql_local}",
                            "rm -rf {$stock} pre-vs"
                        ]);

                        $this->mySQL($sql_local);
	                    $wpdb->query("DELETE FROM {$new_table_prefix}users WHERE ID=1");
	                    $wpdb->query("INSERT INTO {$new_table_prefix}users SELECT * FROM {$old_table_prefix}users");

                    } catch (\Exception $e){
                        $splash = '<style> body, html { height: 100%; }
    body { 
	   background-image: url("../wp-content/plugins/soup-stock-release/img/vacationsoup-map-bg.jpg");
       background-position: center;
       background-repeat: no-repeat;
       background-size: cover;
       max-width: 100%;
       font-family: "Open Sans", sans-serif;
    }
    #content {
       width: 60%;
       margin: 0 auto;
       background: rgba(255,255,255,0.8);
       padding: 20px;
    }
}
			                        </style>';

                        throw new \Exception("{$splash}<div id='content'>
                    <h1>Unrecoverable error during installation</h1>
                    <p>This Wordpress is now likely to be corrupted, a new Installation is necessary.</p>".$e->getMessage().'</div>' );
                    }
                }
            }
        } catch (\Exception $e){
            wp_die("<h1>Vacation Soup Stock Installation Failed</h1><br />".$e->getMessage());
        }
        header('Location: /');
    }

    private function pathfinder($cmd){
	    foreach (['/usr/local/bin/','/usr/bin/','/bin/','/Applications/MAMP/Library/bin/'] as $path){
		    if (file_exists($path.$cmd)) {
			    return $path.$cmd;
		    }
	    }
	    return '/bin/fail &&';
    }

    /**
     * @param $cmds string[] The array of shell commands to run
     * @return string[] The Array of results
     * @throws \Exception if command fails
     */
    function exec_array($cmds) {
        foreach ($cmds as $command){
            $return = $output = 'Not run';
            $when = date("D M j G:i:s T Y");
            $ex = 'cd '.ABSPATH.' && '.$command.' 2>&1';
            exec($ex, $output, $return);
            $this->results[] = compact('when','command','ex','output','return');
            if (0!=$return){ // Oops
                throw new \Exception("Error running shell command: ".$command."<pre>\n".print_r($this->results,1)."</pre>");
            }
        }
    }


    /**
     * Throw up if this is not a plain new installation
     */
    private function checkForNewInstallation(){
        global $table_prefix;
        if (!empty($when = get_option('vs-released'))){
            throw new \Exception("Vacation Soup Stock Installer was already run. Installed version from ".$when);
        }
        // There must only be 1 plugin installed (this one)
        if (count(get_option('active_plugins'))>3){
            throw new \Exception("The only active plugins must be Vacation Soup Stock and the SiteGround defaults of Jetpack and SG Optimizer. If your hosting provider automatically installs a plugin on a new installation, it must be deactivated.");
        }
        // There must be only 1 user created
        if (count_users()['total_users'] != 1){
            throw new \Exception("The only user must be the one you created on installation. If your hosting provider automatically adds a user, it must be deleted. There are ".print_r(count_users(),true)." users.");
        }
        // There must be only 1 post created
        if (wp_count_posts()!=1){
            throw new \Exception("There can only be 1 example post on installation. If your hosting provider automatically adds other posts, they must be deleted.");
        }
        // The existing database tables cannot be prefixed vsoup_
        if (self::RELEASE_PFX==$table_prefix){
            throw new \Exception("Your wordpress is configured with a table prefix of '".self::RELEASE_PFX."' which is reserved for Vacation Soup.");
        }
        // The domain must not be where we came from
        if ($_SERVER['HTTP_HOST'] == self::RELEASE_DOMAIN) {
        	throw new \Exception("This plugin is running on a Vacation Soup host, which is not allowed, it must be run on client installations.");
        }
        // The new domain must be https
        $isSecure = false;
        if (isset($_SERVER['HTTPS']) && $_SERVER['HTTPS'] == 'on') {
            $isSecure = true;
        }
        // These checks are for installations behind a load balancer that decodes https for you
        elseif (!empty($_SERVER['HTTP_X_FORWARDED_PROTO']) && $_SERVER['HTTP_X_FORWARDED_PROTO'] == 'https' || !empty($_SERVER['HTTP_X_FORWARDED_SSL']) && $_SERVER['HTTP_X_FORWARDED_SSL'] == 'on') {
            $isSecure = true;
        }
        if (!$isSecure){
            throw new \Exception("This site is not secured by https. You must configure https (also known as SSL) for this website before you install Vacation Soup.");
        }
    }

    private function mySQL($filename){
        GLOBAL $wpdb;
        $progressFilename = $filename.'_filepointer'; // tmp file for progress
        $errorFilename = $filename.'_error'; // tmp file for erro

        ($fp = fopen($filename, 'r')) OR die('failed to open file:'.$filename);

        $queryCount = 0;
        $query = '';
        while( ($line=fgets($fp, 8192000)) ){ // removed: $deadline>time() AND
            set_time_limit(30);
            if(substr($line,0,2)=='--' OR trim($line)=='' ){
                continue;
            }

            $query .= $line;
            if( substr(trim($query),-1)==';' ){
                if( false === $wpdb->query($query) ){
                    $error = 'Error performing query \'<strong>' . $query . '\': ' . $wpdb->last_error;
                    file_put_contents($errorFilename, $error."\n");
                    wp_die($error);
                }
                $query = '';
                file_put_contents($progressFilename, ftell($fp)); // save the current file position for
                $queryCount++;
            }
            $when = date("D M j G:i:s T Y");
            $command = substr($query,0,80).(strlen($query)>80)?'...':'';
            $output = [
                $queryCount.' queries processed',
                ftell($fp).'/'.filesize($filename).' '.(round(ftell($fp)/filesize($filename), 2)*100).'%'."\n"
            ];

            $this->results[] = compact('when','command','output');
        }
    }
    public function showHoldingPage(){
    	?><!DOCTYPE html>
	    <!--[if IE 8]>
	    <html xmlns="http://www.w3.org/1999/xhtml" class="ie8 wp-toolbar"  lang="en-US">
	    <![endif]-->
	    <!--[if !(IE 8) ]><!-->
	    <html xmlns="http://www.w3.org/1999/xhtml" class="wp-toolbar"  lang="en-US">
	    <!--<![endif]-->
	    <head>
		    <meta http-equiv="Content-Type" content="text/html; charset=UTF-8" />
		    <title>Installing Vacation Soup Stock</title>
		    <style type="text/css">
			    body, html {
				    height: 100%;
			    }

			    .bg {
				    background-image: url("../wp-content/plugins/soup-stock-release/img/vacationsoup-map-bg.jpg");
				    height: 100%;
				    background-position: center;
				    background-repeat: no-repeat;
				    background-size: cover;

				    font-family: "Open Sans", sans-serif;
			    }
			    #content {
				    width: 60%;
				    margin: 15% 0 0 20%;
				    text-align: center;
			    }
			    #content IMG {
				    display: block;
				    min-width: 10%;
                    max-width: 100%;
				    margin: auto;
			    }
		    </style>
	    </head>
	    <body class="bg">
	    <div id="content">
            <img src="../wp-content/plugins/soup-stock-release/img/vacation-soup-logo.png" />
		    <h1>Installing Vacation Soup Stock</h1>
		    <img src="../wp-content/plugins/soup-stock-release/img/loading.svg" />
		    <h3>You will be redirected to your login screen when this is complete. Do not refresh your screen, it should take less than a minute.</h3>
        </div>
	    </body>
        <script>
            window.location.replace('<?= $_SERVER['REQUEST_URI'] ?>?splash&_holding');
        </script>
    </html>

		<?php
		ob_end_flush();
		flush();
    }
}